lib/teleterm/services/desktop/directorysharing.go

```diff
@@ -0,0 +1,296 @@
+	FileTypeDir
+)
+
+const StandardDirSize = 4096
+
+func (s *DirectoryAccess) Open(baseDir string) error {
+	if s.safePathGetter != nil {
+		return errors.New("only one shared directory is supported")
+	}
+
+	base, err := filepath.EvalSymlinks(baseDir)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	stat, err := os.Stat(base)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	if !stat.IsDir() {
+		return trace.BadParameter("%q is not a directory", baseDir)
+	}
+
+	s.safePathGetter = func(relativePath string) (string, error) {
+		full := filepath.Join(base, relativePath)
+		resolved, err := filepath.EvalSymlinks(full)
+		if err != nil {
+			// EvalSymlinks returns an error if the target file does not exist.
+			// In that case, attempt to resolve the symlinks of the parent directory instead.
+			if os.IsNotExist(err) {
+				parent := filepath.Dir(full)
+				resolvedParent, perr := filepath.EvalSymlinks(parent)
+				if perr != nil {
+					return "", trace.Wrap(perr)
+				}
+
+				// Reconstruct the full path by joining the resolved parent with the original file name.
+				resolved = filepath.Join(resolvedParent, filepath.Base(full))
+			} else {
+				return "", trace.Wrap(err)
+			}
+		}
+		if !isSubPath(base, resolved) {
+			return "", trace.BadParameter("path escapes from parent")
+		}
+		return resolved, nil
+	}
+	return nil
+}
+
+func isSubPath(parent, child string) bool {
+	return child == parent || strings.HasPrefix(child, parent+string(filepath.Separator))
+}
+
+// Stat retrieves metadata about a file or directory at the given path.
+func (s *DirectoryAccess) Stat(relativePath string) (*FileOrDirInfo, error) {
+	path, err := s.safePathGetter(relativePath)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	stat, err := os.Stat(path)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	info, err := s.readFileOrDirInfo(relativePath, stat)
+	return info, trace.Wrap(err)
+}
+
+// ReadDir lists files and directories within the given directory path, skips symlinks.
+func (s *DirectoryAccess) ReadDir(relativePath string) ([]*FileOrDirInfo, error) {
+	path, err := s.safePathGetter(relativePath)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	entries, err := os.ReadDir(path)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	var results []*FileOrDirInfo
+	for _, entry := range entries {
+		fileInfo, err := entry.Info()
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		// Skip symlinks, we can't present them properly in the remote machine.
+		if fileInfo.Mode().Type()&os.ModeSymlink != 0 {
+			continue
+		}
+
+		entryRelativePath := filepath.Join(relativePath, fileInfo.Name())
+		fileOrDir, err := s.readFileOrDirInfo(entryRelativePath, fileInfo)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		results = append(results, fileOrDir)
+	}
+
+	return results, nil
+}
+
+// Read reads a slice of a file.
+func (s *DirectoryAccess) Read(relativePath string, offset int64, length uint32) ([]byte, error) {
```

Comment:
Probably not a huge deal, but by having us return a slice that means we need to make an allocation for each read operation, which might hurt performance when reading large files.

Do we have control over this interface so that we can make it accept a `[]byte` to read into?

Reply:
Yeah, we can control the interface, the method now accepts `[]byte`.